src/omv/sensor.c

```diff
@@ -1125,73 +1193,136 @@ int sensor_snapshot(sensor_t *sensor, image_t *image, streaming_cb_t streaming_c
             return -1;
     }
 
-    if (streaming_cb) {
-        image->pixels = NULL;
-    }
-
     // If two frames fit in ram, use double buffering in streaming mode.
-    doublebuf = ((length*2) <= OMV_RAW_BUF_SIZE);
+    bool doublebuf = ((length*2) <= OMV_RAW_BUF_SIZE);
 
     do {
-        // Clear line counter
+        // Clear the line counter variable before we allow more data to be received.
         line = 0;
 
-        // Snapshot start tick
-        tick_start = HAL_GetTick();
-
-        // Enable DMA IRQ
-        HAL_NVIC_EnableIRQ(DMA2_Stream1_IRQn);
+        // If DCMI_DMAConvCpltUser() happens before waiting_for_data = true; below then the
+        // transfer is stopped and it will be re-enabled again right afterwards. We know the
+        // transfer was stopped by checking DCMI_CR_ENABLE.
+
+        waiting_for_data = true;
+
+        // We will be in one of the following states now:
+        // 1. No transfer is currently running right now and DCMI_CR_ENABLE is not set.
+        // 2. A transfer is running and we are waiting for the data to be received.
+
+        // We are not using DCMI_CR_CAPTURE because while this bit when cleared stops the capture...
+        // It does not actually go low. DCMI_CR_ENABLE stops the capture when cleared and stays low.
+        //
+        // When DCMI_CR_ENABLE is cleared during a DCMI transfer the hardware will automatically
+        // wait for the start of the next frame when it's re-enabled again below. So, we do not
+        // need to wait till there's no frame happening before enabling.
+        if (!(DCMI->CR & DCMI_CR_ENABLE)) {
+            // Note that HAL_DCMI_Start_DMA and HAL_DCMI_Start_DMA_MB are effectively the same
```

Comment:
I'd remove this whole comment just to make things more readable, it doesn't add anything important.

Reply:
It's useful.